qemu: should only warn about KVM on Linux

mschwaig opened the pull request:
PR #125948 introduced a warning when KVM is not avialalbe,
which should only be emitted on Linux.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Inaccurate warning.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

alyssais commented:
I think it would be better to not add the makeWrapper --run at all when !stdenv.isLinux — otherwise we're running a script we know ahead of time will do nothing.

mschwaig commented:
> I think it would be better to not add the makeWrapper --run at all when !stdenv.isLinux — otherwise we're running a script we know ahead of time will do nothing.

You're right. I have fixed this now and made it so the runtime check for KVM only happens on Linux as well.

Unfortunately I cannot test this PR on Darwin.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -242,13 +242,20 @@ stdenv.mkDerivation rec {
+  let
+    copyKvmWarningsPathToOutput = ''
+      install -m755 -D $emitKvmWarningsPath $out/libexec/emit-kvm-warnings
+    '';
```
I think it would be clearer (and more idiomatic) if `copyKvmWarningsPathToOutput` was just inlined with `lib.optionalString`.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -242,13 +242,20 @@ stdenv.mkDerivation rec {
-                  --run $out/libexec/emit-kvm-warnings \
-                  --add-flags "\$([ -r /dev/kvm -a -w /dev/kvm ] && echo -enable-kvm)"
+                  $out/bin/qemu-kvm ${
+                  if stdenv.isLinux then addKvmWarningsToWrapper else ""}
```
We have a helper for this:

```nix
lib.optionalString stdenv.isLinux addKvmWarningsToWrapper
```

mschwaig committed:
qemu: should only enable and warn about KVM on Linux

PR #125948 introduced a warning when KVM is not available,
which should only be emitted on Linux.
Similarly the -enable-kvm flag itself also never
needs to be added on platforms other than Linux.

mschwaig commented:
Thanks for your patience with reviewing this.

I did what you suggested and used `lib.optionalString`.

I inlined the Linux-specific arguments to makeWrapper as well, but I left them inside a template expression because that works better for placing the required newlines after that optional expression.

mschwaig commented:
I think this is now obsolete due to https://github.com/NixOS/nixpkgs/pull/157696

I was initially worried about how that change might affect Non-NixOS Linux systems, but it seems it's fine.

On those distributions Nix has no control over the `kvm` kernel modules and the permissions for `/dev/kvm`, which can make KVM run without hardware acceleration on such systems, even when it's properly installed via distribution mechanisms.

I thought it might make sense to at least keep the warning, but I think that's not necessary.

If I break the permissions for KVM on my local NixOS system
```
$ sudo chmod o-rw /dev/kvm
```
and then launch `qemu-kvm`
```
$ qemu-kvm
Could not access KVM kernel module: Permission denied
qemu-kvm: failed to initialize kvm: Permission denied
qemu-kvm: falling back to tcg
[...]
```
QEMU warns me about the broken state of KVM directly and proceeds without hardware acceleration, which was the whole point of the wrapper. The only benefit of the wrapper is that it makes suggestions how to resolve them, but I don't think that's enough for keeping it.

(I would assume that other Linux Distros behave similarly when KVM is not installed or the permissions are wrong, but I have not tested that.)
